packages/dds/tree/docs/wip/identifiers.md

```diff
@@ -0,0 +1,73 @@
+
+## Why Identifiers
+
+Why use identifiers at all? SharedTree has paths, are not those already sufficient "references" or "handles" to nodes in the tree? No, a node's identifier is related to, but not the same thing as, the path to that node in the tree. A path points to a location within a tree, but it does not necessarily capture any semantic meaning about the identity of the node it points to. Consider a node A which is deleted, and then consider a node B which is inserted in the same place. Node A and Node B have the same path (although they are in different revisions) but they are not the same node; they should therefore not share the same identifier. Likewise, consider a node A which is moved from location X to location Y in the tree. Node A has a different path after the move than it did before, but semantically it is the "same node"; it should have the same identifier in both locations.
+
+## Proposal Overview
+
+Identifiers are stored under nodes in fields. A SharedTree has an optional index, the IdentifierIndex. The IdentifierIndex allows a user to provide an identifier and retrieve a cursor to a node with that identifier. An identifier is a special type (a numerically compressed node identifier). The index maintains a mapping from each identifier to a path. If history is enabled, then each identifier maps to a list of tuples of [path, revision]. The index is parameterized over a field key, e.g. "identifier", and will consider all nodes that have this field to be identified by the data in that field. That data ought to be an identifier; if it is not, the index can either ignore it or fail.
+
+## Updating the Index
+
+Like all indexes, the IdentifierIndex can be serialized and rehydrated. When a SharedTree loads a document, the IdentifierIndex loads its last serialized state. From then on, it is updated as part of every operation:
+
+* Insert: The tree being inserted is fully traversed in a search for identifier fields. Any that are found are added to the index. If other indexes are also traversing the tree for their own purposes, it might make sense to have SharedTree provide a single traversal that the indexes can hook into, rather than having each index do its own walk.
+* Delete: The tree being deleted is fully traversed in a search for identifier fields. Any that are found are removed from the index. For a large, virtualized tree this might require downloading parts of the tree that haven't been loaded yet; therefore, updating the index after a delete becomes an async operation. At first glance, this seems to imply op processing must be async, which is probably not acceptable. There may be alternative strategies; for example, the index could allow the removal of the identifiers to run asynchrounously in the background. Then, for each identifier query, the index validates that the mapping it has is still valid by looking up the node in the forest by its path and then checking that its identifier matches the one in the index.
```

Comment:
Depending on how the index stores information, it may be possible to avoid traversal of deleted subtrees. That's admittedly not something this document should worry about, just mentioning it in case I forget later. Feel free to resolve this comment without further action.